src/importer.rs

```diff
@@ -58,26 +58,81 @@ pub struct Model {
     pub geometry: Geometry,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
 pub struct FileMetadata {
     checksum: u32,
-    last_modified: std::time::SystemTime,
+    last_modified: SystemTime,
 }
 
 pub type ImporterResult = Result<Vec<Model>, ImporterError>;
 
-/// `Importer` takes care of importing of obj files and caching of their
-/// internal representations. It holds paths to files, their metadata
-/// (checksums, timestamps) and models of parsed obj files.
+/// An interface for caching of obj files.
+///
+/// The source is expected to be file with path and FileMetadata.
+pub trait Cache {
+    /// Returns models if modified timestamp of file on given `path` didn't
+    /// change.
+    fn get_if_not_modified(&self, path: &str, modified: SystemTime) -> Option<Vec<Model>>;
+
+    /// Returns models if checksum of source file is already cached.
+    ///
+    /// This means contents of file were cached, however it could have been
+    /// from a different path. Therefore file path and metadata should be
+    /// cached again.
+    fn get_by_checksum(&self, checksum: u32) -> Option<Vec<Model>>;
+
+    /// Sets given data in cache.
+    ///
+    /// Most likely there is going to be multiple caching structures present, so
+    /// this method accepts all available data and picks only whatever makes
+    /// sense to cache.
+    fn set(&mut self, path: String, metadata: FileMetadata, models: &[Model]);
+}
+
+/// Cache with no size limits. Anything set is kept until the application exits.
 #[derive(Debug, Default)]
-pub struct Importer {
+pub struct EndlessCache {
     path_metadata: HashMap<String, FileMetadata>,
     loaded_models: HashMap<u32, Vec<Model>>,
 }
 
-impl Importer {
-    pub fn new() -> Self {
-        Default::default()
+impl Cache for EndlessCache {
+    fn get_if_not_modified(&self, path: &str, modified: SystemTime) -> Option<Vec<Model>> {
+        if let Some(path_metadata) = self.path_metadata.get(path) {
+            if path_metadata.last_modified == modified {
+                return Some(
+                    self.loaded_models
+                        .get(&path_metadata.checksum)
+                        .expect("Should get loaded models by obj file's checksum")
+                        .clone(),
+                );
+            }
+        }
+
+        None
+    }
+
+    fn get_by_checksum(&self, checksum: u32) -> Option<Vec<Model>> {
+        self.loaded_models.get(&checksum).cloned()
+    }
+
+    fn set(&mut self, path: String, metadata: FileMetadata, models: &[Model]) {
+        self.path_metadata.insert(path, metadata);
+        self.loaded_models
+            .entry(metadata.checksum)
+            .or_insert_with(|| models.to_vec());
+    }
+}
+
+/// `Importer` takes care of importing of obj files and caching of their
+/// internal representations.
+pub struct Importer<'a, C> {
+    cache: &'a mut C,
```

Comment:
Hmm, any reason the `cache` is not owned by the importer?

Reply:
It used to be until I started to experiement with home-grown mocking and I wanted to see what happened inside of cache object. Now that we're using library for that, I'll see how it behaves, but I'm pretty sure those mocks will need to be passed by reference to be usable in tests.